utils/python/CIME/XML/grids.py

```diff
@@ -5,25 +5,143 @@
-        self.groups={}
+
+    def get_grid_info(self, name, compset):
+        """
+        Find the matching grid node
+        """
+        nodes = self.get_nodes("grid")
+        gridinfo = {}
+
+        # first search for all grids that have a compset match - if one is found then return
+        for node in nodes:
+            if "compset" in node.attrib:
+                attrib = node.get("compset")
+                compsetRE = re.compile(attrib)
+                compset_match = re.search(attrib,compset)
+                if compset_match is not None:
+                    alias = self.get_value("alias",root=node)
+                    sname = self.get_value("sname",root=node)
+                    lname = self.get_value("lname",root=node)
+                    if alias == name or lname == name or sname == name:
+                        logger.debug("Found node compset match: %s and lname: %s" % (attrib, lname))
+                        component_grids = self._get_component_grids(lname)
+                        domains  = self._get_domains(component_grids)
+                        gridmaps = self._get_gridmaps(component_grids)
+                        gridinfo.update(domains)
+                        gridinfo.update(gridmaps)
+                        gridinfo["GRID"] = lname
+                        return gridinfo
+
+
+        # if no matches were found for a possible compset match, then search for just a grid match with no
+        # compset attribute
+        for node in nodes:
+            if "compset" not in node.attrib:
+                sname = self.get_value("sname",root=node)
+                alias = self.get_value("alias",root=node)
+                lname = self.get_value("lname",root=node)
+                if alias == name or lname == name or sname == name:
+                    logger.debug("Found node compset match: %s and lname: %s" % (attrib, lname))
+                    component_grids = self._get_component_grids(lname)
+                    gridinfo.update(self._get_domains(component_grids))
+                    gridinfo.update(self._get_gridmaps(component_grids))
+                    gridinfo["GRID"] = lname
+                    return gridinfo
+
+        expect (False,
+                "grid '%s'  is not supported, use manage_case to determine supported grids " %name)
+
+    def get_value(self, item, attributes=None, root=None):
+        if root is None:
+            root = self.root
+        node = self.get_optional_node(item,attributes=attributes, root=root)
+        if node is not None:
+            return node.text
 
     def _get_component_grids(self, name):
         gridRE = re.compile(r"[_]{0,1}[a-z]{1,2}%")
-        component_grids = gridRE.split(name.text)[1:]
+        component_grids = gridRE.split(name)[1:]
+
         return component_grids
 
+    def _get_domains(self, component_grids):
+        # use component_grids to create grids dictionary
+        grids = [("ATM", component_grids[0]), \
```

Comment:
Yes -  
I'm afraid that the quest for complete generality breaks down in config_grids.xml.  There is a lot of specific information here about interactions between components that currently is expressed in code instead of in XML.